### PR TITLE
Improve caret translation logic

### PR DESCRIPTION
**Pull request by jag250**

### CODE REVIEW

File: metadata_please/source_checkout.py [hunk 1]
```diff
@@ -139,22 +140,35 @@ def _translate_caret(specifier: str) -> str:
...
+
+            # Set the rest of the values to 0
+            for i in range(2, len(version_parts)):
+                version_parts[i] = 0
+
+        else:
+            # Next patch version
+            version_parts[2] += 1
+
+    incremented_version = ".".join([str(i) for i in version_parts])
```
Comment (thatch) [inline metadata_please/source_checkout.py:177]: This could use a comment about how this is only the numeric parts, any pre- or post- is discarded.

File: metadata_please/source_checkout.py [hunk 2]
```diff
@@ -213,7 +227,26 @@ def from_poetry_checkout(path: Path) -> bytes:
                 optional = False
 
             if not version:
-                # e.g. git, path or url dependencies, skip for now
+                # e.g. git, path or url dependencies
+                if "path" in v:
```
Comment (thatch) [inline metadata_please/source_checkout.py]: Can you split this change out to a separate PR?  This needs more thought and probably some sort of X- prefixed name, as I'm pretty sure they are not valid Requires-Dist.

File: metadata_please/source_checkout.py [hunk 3]
```diff
@@ -139,22 +140,35 @@ def _translate_caret(specifier: str) -> str:
...
-            break
+
+    version = Version(specifier[1:])
+    version_parts = list(version.release)
+    if version.is_prerelease or version.is_postrelease:
+        # Return next version, incrementing the least significant number
+        version_parts[-1] += 1
+
     else:
-        raise ValueError("All components were zero?")
```
Comment (thatch) [inline metadata_please/source_checkout.py:157]: Can you add a test with `^0` and make sure it does something predictable?  I think it's invalid, and I don't know if it will raise (which is fine) or do something less useful like `>=0,<1`

Comment (jag250) [reply metadata_please/source_checkout.py:157]: That case would return `>=0,<1`. Similarly, `^0.0` returns `>=0.0,<1.0` and `^0.0.0` returns `>=0.0.0,<1.0.0`. Is it better to raise?

Comment (thatch) [reply metadata_please/source_checkout.py:157]: There are two examples given at the bottom of https://python-poetry.org/docs/dependency-specification/#caret-requirements that are the best spec I can find.  My personal read of

> An update is allowed if the new version number does not modify the left-most non-zero digit in the major, minor, patch grouping.

is that this isn't a case they've really designed for.  I'd think `^0.0.x` is compatible with 0.0.xpost1 for example, but the prose implies that isn't the case.

I'm just saying, please add tests for the zero case if it returns now instead of raising.

Comment (jag250) [reply metadata_please/source_checkout.py:157]: Added a check for this case that raises a `ValueError`!

**Review by thatch — APPROVED** [review]
ok!
